Impl `from_reader`

wcampbell0x2a opened the pull request:
Builds upon #345 and #342. Truly fixing #44 and #344. And #252, #105. And somewhat #290.

Ok that's enough. This adds `from_reader` which can take in a Read! I also went ahead and implemented some long standing issues with this crates and _actually_ make a `Container` that can properly specialize on aligned byte reads.

There's more when I have time to review, many TODOs are still standing. But the benchmarks look good!

### benchmarks
I updated the benchmarks, so you can't just compare against master:
```
> critcmp before after
group                  after                                  before
-----                  -----                                  ------
deku_read_bits         1.24   845.8±14.29ns        ? ?/sec    1.00   679.5±11.83ns        ? ?/sec
deku_read_byte         1.00     17.8±0.25ns        ? ?/sec    2.12     37.8±3.35ns        ? ?/sec
deku_read_enum         1.00     15.3±0.15ns        ? ?/sec    2.04     31.2±0.81ns        ? ?/sec
deku_read_vec          1.00    676.8±7.04ns        ? ?/sec    2.16  1459.5±40.22ns        ? ?/sec
deku_read_vec_perf     1.00   640.6±34.63ns        ? ?/sec    2.54  1627.4±31.78ns        ? ?/sec
deku_write_bits        1.00    125.3±3.10ns        ? ?/sec    1.04   130.2±11.12ns        ? ?/sec
deku_write_byte        1.00    161.6±4.86ns        ? ?/sec    1.02    165.0±5.91ns        ? ?/sec
deku_write_enum        1.00    105.6±1.06ns        ? ?/sec    1.03    109.0±7.20ns        ? ?/sec
deku_write_vec         1.00      4.6±0.04µs        ? ?/sec    1.06      4.9±0.07µs        ? ?/sec
deku_write_vec_perf    1.00      4.6±0.08µs        ? ?/sec    1.06      4.9±0.06µs        ? ?/sec
```



### todo
- [x] Add example of using new `from_reader` to read from file
- [x] Add MSRV
- [x] Using `--engine = llvm` on tarpulin might produce more accurate coverage results? Update: still inaccurate. https://github.com/taiki-e/cargo-llvm-cov seems to be more accurate. `cargo +nightly llvm-cov --doctests --release --html --ignore-filename-regex attributes` produces very good results.
- [x] Fix/Enhance "logging" feature
- [x] I think I need to keep around the old derives that gives `read(..)` for `no_std` enviroments
- [x] A ton of testing. I'll test out with my own crates and make sure this whole API works.
- [x] Conversation about things like `#[deku(cond = "!deku::rest.is_empty()")]` not working. not sure the solution on this since there is no way of knowing if `Read` is done except trying to read [u8; 1] and seeing it it errors and buffer the results
- [x] BufRead impl should __really__ be encouraged for most big reads. Results may vary but a PERFORMANCE.md might be good.

### Example usage
- UBI: :heavy_check_mark:  https://github.com/wcampbell0x2a/income 
- SquashFS: :heavy_check_mark: https://github.com/wcampbell0x2a/backhand/pull/286 [<img alt="build status" src="https://img.shields.io/github/actions/workflow/status/wcampbell0x2a/backhand/main.yml?branch=deku-update&style=for-the-badge" height="20">](https://github.com/wcampbell0x2a/backhand/actions?query=branch%3Adeku-update)
- WIP: :heavy_check_mark: adsb: https://github.com/rsadsb/adsb_deku/pull/212 [<img alt="build status" src="https://img.shields.io/github/actions/workflow/status/wcampbell0x2a/backhand/main.yml?branch=update-deku&style=for-the-badge" height="20">](https://github.com/wcampbell0x2a/backhand/actions?query=branch%3Aupdate-deku)
- CPIO: :heavy_check_mark: https://github.com/wcampbell0x2a/cpio-deku
- asterix: https://github.com/wcampbell0x2a/asterix-rs/pull/9
- grain of salt benchmarks: https://github.com/wcampbell0x2a/deku-bench/pull/1

sharksforarms commented:
Very excited about this!

wcampbell0x2a commented:
@sharksforarms Do we need to still keep the normal `DekuRead` implementations?

They won't (if this patch is merged) be used in any derived code anymore.

I think that would remove almost all public uses of bitvec. I think my container/reader could just hand them back an [bool; 8] for trailing bits after from_bytes or something. I haven't worked out that API usability yet.

Also, I think I'll rename container to reader.

wcampbell0x2a commented:
@sharksforarms `from_bytes` could be changed to 'from_reader'. But `from_reader` is already taken (for the with ctx case). And ideas on function name? Or change the other function? I guess changing the other to `from_reader_ctx` is my idea.

sharksforarms commented:
> @sharksforarms `from_bytes` could be changed to 'from_reader'. But `from_reader` is already taken (for the with ctx case). And ideas on function name? Or change the other function? I guess changing the other to `from_reader_ctx` is my idea.

SGTM, let's see what it looks like with that

sharksforarms reviewed (commented):
Looks great! The PR is mostly big as it updates many tests and areas.

It would be great to update the changelog too with a migration guide, maybe there's something I missed.

I was also wondering if it'd be possible to remain backwards compat by emitting a `from_bytes` with a `#[deprecated]` which just calls the reader api under the hood? What do you think?

 I left a couple of questions/comments.

wcampbell0x2a commented on src/reader.rs:
```diff
@@ -0,0 +1,276 @@
+pub enum ReaderRet {
+    /// Successfully read bytes
+    Bytes,
+    /// Read Bits intead
```
fix spelling

wcampbell0x2a commented on tests/test_alloc.rs:
```diff
@@ -57,10 +57,10 @@ mod tests {
             })
             .0,
-            (4, 0, 4)
+            (5, 0, 5)
```
This should come from [this](https://github.com/sharksforarms/deku/pull/352/files#diff-267c9f4da66412a9f439ac08d224356fe24265b5e1cebb6c44c2d55b96414513R177). We basically need to store the bits that weren't used into `reader.leftover`. This currently allocates. 

__Although__ we should be able to just keep around a BitArray of [bool; 8] or something like that and store them just in the struct since I don't think it would ever keep around more then 8. It's a good TODO for later, especially as it most likely will remove most perf regressions for reading only bits.

wcampbell0x2a commented on README.md:
````diff
@@ -30,6 +30,8 @@ no_std:
 deku = { version = "0.16", default-features = false, features = ["alloc"] }
 ```
 
+*Compiler support: requires rustc 1.65+*
````
Yep! I think the world has moved on from actions-rs, so more updating to the CI is in due.

wcampbell0x2a commented:
Github seems to have removed conversations when I forced pushed after rebasing from master -_-

Luckily I have the emails:

> I was also wondering if it'd be possible to remain backwards compat by emitting a from_bytes with a #[deprecated] which just calls the reader api under the hood? What do you think?

I'm not sure this is possible, I think the lifetime of the input `(&'a [u8)` prevents this, atleast from my small testing.

> It would be great to update the changelog too with a migration guide, maybe there's something I missed.

Will do!

> are we worried about crate maintenance here?

> after doing a quick search it doesn't seem like there's a "goto" crate for no_std io, I did find https://github.com/no-std-io/no-std-io which is semi-recently updated, wdyt?

I switched to this one. At least it's published to crates.io.

wcampbell0x2a commented:
@sharksforarms I updated the CHANGELOG, it's not my strong suit so lmk what you think of it.

wcampbell0x2a commented:
@sharksforarms I misspoke previously and said that `from_bytes(..)` wasn't possible anymore. It can be added back with the same signature :confetti_ball: which should ease API adoption.

I added it with this commit: https://github.com/sharksforarms/deku/pull/352/commits/9c75e27d6f1aed3a1f8118de5ec3870ec2b5d3b7.

I'll add it back into the API docs and update the CHANGELOG. (and test it more)

wcampbell0x2a commented:
> @sharksforarms I misspoke previously and said that `from_bytes(..)` wasn't possible anymore. It can be added back with the same signature 🎊 which should ease API adoption.
> 
> I added it with this commit: [9c75e27](https://github.com/sharksforarms/deku/commit/9c75e27d6f1aed3a1f8118de5ec3870ec2b5d3b7).
> 
> I'll add it back into the API docs and update the CHANGELOG. (and test it more)

The documentation here relies more on your own opinions. Should I replace all the docs that I replaced back with `from_bytes`, or should I keep then `from_reader`? 

 I think in `src/lib.rs` I'll reset the first example back to `from_bytes` and add a overview of `from_bytes` and `from_reader`.

sharksforarms commented:
> The documentation here relies more on your own opinions. Should I replace all the docs that I replaced back with `from_bytes`, or should I keep then `from_reader`? 
> 
>  I think in `src/lib.rs` I'll reset the first example back to `from_bytes` and add a overview of `from_bytes` and `from_reader`.

Yeah I think having an example for both makes sense!

wcampbell0x2a commented on CHANGELOG.md:
````diff
@@ -2,6 +2,142 @@
+```
+
+- The more internal (with context) `read(..)` was replaced with `from_reader_with_ctx(..)`.
+Now unused variables `deku::input`, `deku::input_bits`, and `deku::rest` were removed. `deku::reader` is the replacement.
````
Pretty much.

deku::input was removed, as we don't cache the bytes read from `reader`. This could be added in, but I didn't want to add more to that hot loop.

I used that in one of my crates, and I just replaced it with a wrapper around the `read(..)`: https://github.com/rsadsb/adsb_deku/pull/212/files#diff-2149937cf9daa006df48008e1a47b5fac20f39821233df371e60e75c1abd4a66R158.

I'm not against adding a `deku::input` back in with the cache, as long as perf doesn't suffer like it did when I first tested it.

sharksforarms replied:
Ok, would it be possible to provide an example in the examples folder and update the changelog to point to the alternative? This may avoid people creating an issue :sweat_smile:

wcampbell0x2a commented on CHANGELOG.md:
```diff
@@ -2,6 +2,142 @@
+
+let (amt_read, ret_read) = TestDeku::from_reader((&mut c, 0)).unwrap();
+assert_eq!(TestDeku(0b0110), ret_read);
+assert_eq!(amt_read, 4);
```
I could change it to `(bits_read, ret_read) = TestDeku::from_reader((&mut c, 0)).unwrap();`?

If we change this, we should change `from_bytes`.

wcampbell0x2a commented on CHANGELOG.md:
````diff
@@ -2,6 +2,142 @@
+}
+```
+
+- With the addition of using `Read`, containing a byte slice with a reference is not supported:
````
For sure, i'll add a comment.

wcampbell0x2a replied:
Interestingly, the following works:
```rs
#[derive(Debug)]
struct Nice<'a> {
    buf: &'a [u8]
}

impl<'a> Nice<'a> {
    pub fn from_bytes(mut reader: impl std::io::Read, storage: &'a mut [u8]) -> Nice<'a> {
        reader.read_exact(storage);
       
        Nice {
            buf: storage,
        }
    }
}

fn main() {
    let mut buf = std::io::Cursor::new([0x00, 0x01, 0x02]);
    let mut storage = [0; 3];
    let nice = Nice::from_bytes(&mut buf, &mut storage);
    dbg!(nice);
}
```

I haven't tried it at the scale of deku..

wcampbell0x2a commented on CHANGELOG.md:
```diff
@@ -2,6 +2,142 @@
 
+## Changes
+[#352](https://github.com/sharksforarms/deku/pull/352) replaced the byte slice interface with new interfaces depending on `io::Read`.
+This brings massive performance and usability improvements.
```
I can make an entry in the changelog before we release giving the most up to date benchmarks from that repo.

wcampbell0x2a commented on CHANGELOG.md:
````diff
@@ -2,6 +2,142 @@
+```
+
+- `id_pat` is now required to be the same type as stored id.
+This also disallows using tuples for storing the id:
````
See comments:

old:
```rs
   impl ::deku::DekuRead<'_, ()> for DekuTest {
        fn read(
            __deku_input_bits: &'_ ::deku::bitvec::BitSlice<u8, ::deku::bitvec::Msb0>,
            _: (),
        ) -> core::result::Result<
            (&'_ ::deku::bitvec::BitSlice<u8, ::deku::bitvec::Msb0>, Self),
            ::deku::DekuError,
        > {
            use core::convert::TryFrom;
            let mut __deku_rest = __deku_input_bits;
            // in this version, we read from __deku_rest, but don't use __deku_new_rest,
            // so we don't move the "reader"
            let (__deku_new_rest, __deku_variant_id) = <u8>::read(__deku_rest, ())?;
            let __deku_value = match &__deku_variant_id {
                _ => {
                    let __deku_field_0 = {
                        let (__deku_new_rest, __deku_value) = <(
                            u8,
                            u8,
                        ) as ::deku::DekuRead<'_, _>>::read(__deku_rest, ())?;
                        let __deku_value: (u8, u8) = core::result::Result::<
                            _,
                            ::deku::DekuError,
                        >::Ok(__deku_value)?;
                        __deku_rest = __deku_new_rest;
                        __deku_value
                    };
                    let field_0 = &__deku_field_0;
                    Self::VariantC(__deku_field_0)
                }
            };
            Ok((__deku_rest, __deku_value))
        }
    }
```
new:
```rs
   impl ::deku::DekuReader<'_, ()> for DekuTest {
        fn from_reader_with_ctx<R: ::deku::no_std_io::Read>(
            __deku_reader: &mut ::deku::reader::Reader<R>,
            _: (),
        ) -> core::result::Result<Self, ::deku::DekuError> {
            use core::convert::TryFrom;
            let __deku_variant_id = <u8>::from_reader_with_ctx(__deku_reader, ())?;
            let __deku_value = match &__deku_variant_id {
                _ => {
                    let __deku_field_0 = {
                        // we just use the __deku_variant_id
                        let __deku_value = __deku_variant_id;
                        let __deku_value: (u8, u8) = core::result::Result::<
                            _,
                            ::deku::DekuError,
                        >::Ok(__deku_value)?;
                        __deku_value
                    };
                    let field_0 = &__deku_field_0;
                    Self::VariantC(__deku_field_0)
                }
            };
            Ok(__deku_value)
        }
    }
```

I can't clone the `reader`, and caching the exact `

wcampbell0x2a replied:
This becomes an easy problem in the case of adding the `Seek` trait on reader, hard if not. As we need to create the exact state of `reader` before reading `__deku_variant_id`.

sharksforarms reviewed (approved):
Thanks @wcampbell0x2a ! Sorry for the delayed review, I don't have too much time for open source these days. Feel free to merge and `@` me for more reviews!

wcampbell0x2a commented:
> Thanks @wcampbell0x2a ! Sorry for the delayed review, I don't have too much time for open source these days. Feel free to merge and `@` me for more reviews!

@sharksforarms make some last changes, should be ready to merge

sharksforarms reviewed (approved):
Thanks for all your efforts here! Feel free to merge.

sharksforarms reviewed (dismissed):
Rebased to fix merge conflict

wcampbell0x2a commented:
> Rebased to fix merge conflict

Sweet. I'll squash and make a nice commit msg then merge.

wcampbell0x2a commented on src/lib.rs:
````diff
@@ -98,6 +104,7 @@ If the length of Vec changes, the original field specified in `count` will not g
 Calling `.update()` can be used to "update" the field!
 
 ```rust
+# use std::io::Cursor;
````
Unused

wcampbell0x2a commented on src/lib.rs:
````diff
@@ -57,9 +61,11 @@ assert_eq!(vec![0b0110_1001, 0xC0, 0xFE], data_out);
+can be derived by using the `DekuRead` and `DekuWrite` Derive macros.
 
 ```rust
+# use std::io::Cursor;
````
Unused

wcampbell0x2a commented:
@sharksforarms can only merge with 1 approving review

sharksforarms reviewed (approved):
Thanks feel free to merge! I propose we do squash-merge via github in the future, makes it easier to review as a rebase+force-push could introduce some unreviewed code, I'll change the project setting
